### DISCUSSION TITLE
Higher level Pythonic DOM interface

### DISCUSSION THREAD

**Discussion by fpliger** [pyscript/pyscript#1091] [Proposals]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other feature requests and couldn't find a duplicate (including also the ``type-feature`` tag)
- [X] I confirmed that it's not related to another project area (see the above section)

### What is the idea?

PyScript should provide a pythonic web/dom interface. I'd like to start the discussion here.... 

Later I'll propose what I've been thinking so far, in  the comments, but **if you had the opportunity, how would you design a pythonic `dom` api?**

### Why is this needed

`Pyodide` provides a great `js` interface that allows users to use the Javascript DOM interface directly. It's not very Pythonic though.

We currently have the [not so great] `Element` class in `pyscript.py` that provides some convenience over that `js` API it's not really a well thought intentional API. 

### What should happen?

We should discuss here and reach a higher level agreement on how this API should look like

**Comment by FabioRosado** [comment]
I've moved this into the discussions, hopefully, that's okay 😄 

I've been thinking about this for a while, but I missed the meeting with Antonio and Nicholas where you discussed this topic. I'll add my thoughts/proposal and see what folks think 😄 

## Selecting elements

Currently, you can use the `Element` API to select elements, but only through their ID. Ideally, we would have a QuerySelector class that would allow us to query elements by ID, and class and offer a more advanced selection. This would then return an `Element` or a `List[Element]` that matches the selection. For example:

```python
by_id = FindElement("#test") # returns Element with id 'test'
by_class = FindElement(".test") # Returns list of Element that contains class 'test'
advanced = FindElement("#test > div") # Retuns child node div from parent #test
```

## Creating Elements

We should allow users to create elements easily. I've seen folks using a few different ways:

- Using pure JS to do such
- adding a div with an id and then using `Element` to write to `innerHTML`
- adding a div with an id and then using js to get the element and adding to `innerHTML`

Following the same pattern above, we can create a class that allows users to create elements. For example

```python
create_div = CreateElement('div')
div_with_id = CreateElement('div', id="my-div")
div_wit_class = CreateElement('div', class="my-class")
```

We can then support different methods like:

```python
div = CreateElement('div' class="my-class", id="my-div")
p = CreateElement('p', id="hi").innerText("Hello world!")
div.append(p)
```

Which would be the same as:

```html
<div id="my-div" class="my-class">
  <p id="hi">Hello world!</p>
```

## The Element class

The element class would be the wrapper class for any element in the page, so it would have to accommodate some logic for the `CreateElement` such as `append` to add a childNode. 

It also needs to allow us to instantiate an Element with other things, not just an id. Ideally, we would instantiate `Element` with the respective dom element.

We can then tidy it up a bit and expand the class to allow us to do things such as:

- add event listeners leveraging js (or pyodide, but perhaps we don't want to be too interpreter specific)
- allow basic actions/events such as `click`, `check`, `press` (keypress combo), `type` (typing/textinput), `hover`,  etc


## Closing thoughts

This isn't a rewrite of the API we have. In fact, we already have a function that allows us to [create](https://github.com/pyscript/pyscript/blob/main/pyscriptjs/src/python/pyscript.py#L319-L324) and query elements (although by id only). Perhaps it makes sense start by expanding, reworking the elements we have and see how folks get on? 

Oh and add a bunch of docs and tutorial on how to leverage these new api elements of course :D

**Reply by ramalho** [reply]
Stupid question: why don't we emulate the API of [`ElementTree`](https://docs.python.org/3/library/xml.etree.elementtree.html) from the Python Standard Library, which is also the core API of the awesome `lxml` package?

Python also has a [`xml.dom`](https://docs.python.org/3/library/xml.dom.html) package, but `ElementTree` was created precisely to provide a more Pythonic API.

**Reply by fpliger** [reply]
Nice @ramalho , that's exactly what I was looking into as foundation for this web API in PyScript. During the last chat with @ntoll  and  @antocuni, (I think) Antonio brought up "why don't we just use `xml.dom`" and my answer was that I don't really love that API and didn't find it very pythonic when I had to use it. Over the break was looking at `ElementTree` though and I think it could be a great foundation.

**Comment by ntoll** [comment]
Hi... so PolyPlug has a nascent version of this already: https://github.com/pyscript/polyplug/blob/main/polyplug.py

**Reply by FabioRosado** [reply]
Ah that's great to know thanks.  Since I wasn't in the meeting, was it decided we would use polyplug implementation?
Are we moving the code from polyplug into pyscript?

**Reply by ntoll** [reply]
PolyPlug is separate to Pyscript but used by MicroPython (in PyScript) to interact with the DOM. I feel it important that we retain this clear separation of concerns simply so we're able to keep things clean. ;-)

**Reply by FabioRosado** [reply]
That makes sense, but how are we planning to incorporate the two together?
Will it be a dependency that we have to include in our end so we can import it with `from pyscript import web` or similar?

**Reply by ntoll** [reply]
These are all details to work out... but I hope to show/tell in tomorrow's PyScript FUN session. ;-)

**Reply by FabioRosado** [reply]
Awsome! I'm very excited to see it then 😄

**Comment by fpliger** [comment]
Following up on the comment above. I'm +1 on @ramalho  proposal to use/emulate the API of [ElementTree](https://docs.python.org/3/library/xml.etree.elementtree.html) from the Python Standard Library and use that as a foundation for `Element`.

Tbh, think that this could be a great starting point. On top of that, I'd like to explore an "experimental" API on top of that since I don't really love the `CreateElement` factory approach as I find it verbose and a bit cumbersome (while, to be fair, it may be more semantically explicit(?) That's why I'd like to explore the route and check feedback, because it might be just me lol). For instance, here is some very scrappy braindump:

```python
mydiv = web.div("div1", classes=["pretty", "highlight"]) # --> returns Element
par = web.p("par1", html="Hi!")

# appends par as child
mydiv.append(par)

# fully replaces mydiv.innerHTML
mydiv.write(par)

web.document.append(mydiv)

plot = create_matplotlib_plot(....)
mydiv.append(plot)

# queries mydiv children looking for specific elements/patterns
# Note on the below: query seems more "appropriate" but find seems more user friendly(??)
mydiv.find("p") # --> returns [par]

# -----
el = Element("myid")
el.write("Hi Mom!")

el.write(web.escape("My results are a > 2"))

div = web.query("myid")
div.add_classes("highlight")
# or maybe
div.classes += "highlight"
```

**Comment by pauleveritt** [comment]
I realize there's a lot of momentum towards a DOM-like API. At the same time, the frontend world has spent a few thousand person-centuries fleeing that kind of imperative style. 😀 Just curious, is this the API you expect civilians to use, or behind-the-scenes machinery for something higher-level?

**Reply by fpliger** [reply]
If you are talking about the lower level mapping to JS api, no. That's not the API I expect to be used by most of PyScript users. I think the purpose of this API is mostly a behind-the-scenes for something higher level or just a JS compat sort of layer. (also to keep in mind that the JS Web API are massive and have a cost)

If you are talking about any other ideas in this discussion my answer is also not a yes. When it comes to higher level APIs, I prefer to have a process that enables multiple proposals/paradigms to bloom and the community/users will pick the most popular ones that we should be promoting closer and closer to pyscript "stdlib".

**Comment by pauleveritt** [comment]
Now to answer "if you had the opportunity, how would you design"...in [this step of my writeup](https://pyodide-components.readthedocs.io/en/latest/component_instances.html) one can find my answer:

1. A Python component returns a string of HTML.
2. The JS side gets the string, matches it to the JS "instance" (data-cuid or something) for that component instance.
3. [Idiomorph from HTMX](https://github.com/bigskysoftware/idiomorph) does the DOM patching.

My writeup presumes some component model promises that you won't want:

- A way to keep Custom Element instances mirrored on the JS side with the Python side
- A weirdo-approach to event dispatching

Still, bringing something that feels like an HTML string into the mix, seems pretty sweet. Python had [imperative HTML libraries from 1994](https://pypi.org/project/htmlgen/0.1/) forward and that style didn't really catch on IMO.

**Comment by fpliger** [comment]
> Now to answer "if you had the opportunity, how would you design"...in [this step of my writeup](https://pyodide-components.readthedocs.io/en/latest/component_instances.html) one can find my answer:
> 
> A Python component returns a string of HTML.
> The JS side gets the string, matches it to the JS "instance" (data-cuid or something) for that component instance.
> [Idiomorph from HTMX](https://github.com/bigskysoftware/idiomorph) does the DOM patching.
> 

👍 . Besides Idiomorph patching the DOM, that doesn't seem that far from what Nicholas and I have been talking as well..

>  My writeup presumes some component model promises that you won't want:

> A way to keep Custom Element instances mirrored on the JS side with the Python side
> A weirdo-approach to event dispatching

Not sure we are aligned? What parts you think I don't want?

> Still, bringing something that feels like an HTML string into the mix, seems pretty sweet. Python had [imperative HTML libraries from 1994](https://pypi.org/project/htmlgen/0.1/) forward and that style didn't really catch on IMO.

As mentioned above, I think that the more our users drive the APIs that "win", the better... There are many cases where internal needs that may drive APIs but the more that comes from the community/users the better.

**Reply by pauleveritt** [reply]
> > Now to answer "if you had the opportunity, how would you design"...in [this step of my writeup](https://pyodide-components.readthedocs.io/en/latest/component_instances.html) one can find my answer:
> > A Python component returns a string of HTML.
> > The JS side gets the string, matches it to the JS "instance" (data-cuid or something) for that component instance.
> > [Idiomorph from HTMX](https://github.com/bigskysoftware/idiomorph) does the DOM patching.
> 
> 👍 . Besides Idiomorph patching the DOM, that doesn't seem that far from what Nicholas and I have been talking as well..
> 
> > My writeup presumes some component model promises that you won't want:
> 
> > A way to keep Custom Element instances mirrored on the JS side with the Python side
> > A weirdo-approach to event dispatching
> 
> Not sure we are aligned? What parts you think I don't want?

This topic is limited to just the DOM-like thing. Doesn't need to bring in unrelated stuff from my (now-dead) experiment.
